x-pack/filebeat/module/googlecloud/firewall/config/pipeline.js

```diff
@@ -0,0 +1,331 @@
+}
+
+// PipelineBuilder to aid debugging of pipelines during development.
+function PipelineBuilder(pipelineName, debug) {
+    this.pipeline = new processor.Chain();
+    this.add = function (processor) {
+        this.pipeline = this.pipeline.Add(processor);
+    };
+    this.Add = function (name, processor) {
+        this.add(processor);
+        if (debug) {
+            this.add(makeLogEvent("after " + pipelineName + "/" + name));
+        }
+    };
+    this.Build = function () {
+        if (debug) {
+            this.add(makeLogEvent(pipelineName + "processing done"));
+        }
+        return this.pipeline.Build();
+    };
+    if (debug) {
+        this.add(makeLogEvent(pipelineName + ": begin processing event"));
+    }
+}
+
+function FirewallProcessor(keep_original_message, debug) {
+    var builder = new PipelineBuilder("firewall", debug);
+
+    // The pub/sub input writes the Stackdriver LogEntry object into the message
+    // field. The message needs decoded as JSON.
+    builder.Add("decodeJson", new processor.DecodeJSONFields({
+        fields: ["message"],
+        target: "json"
+    }));
+
+    // Set @timestamp to the LogEntry's timestamp.
+    builder.Add("parseTimestamp", new processor.Timestamp({
+        field: "json.timestamp",
+        timezone: "UTC",
+        layouts: ["2006-01-02T15:04:05.999999999Z07:00"],
+        tests: ["2019-06-14T03:50:10.845445834Z"],
+        ignore_missing: true
+    }));
+
+    if (keep_original_message) {
+        builder.Add("saveOriginalMessage", new processor.Convert({
+            fields: [
+                {from: "message", to: "event.original"}
+            ],
+            mode: "rename"
+        }));
+    }
+
+    builder.Add("dropPubSubFields", function(evt) {
+        evt.Delete("message");
+        evt.Delete("labels");
+    });
+
+    builder.Add("categorizeEvent", new processor.AddFields({
+        target: "event",
+        fields: {
+            category: "firewall-rule",
+            type: "firewall"
+        },
+    }));
+
+    builder.Add("saveMetadata", new processor.Convert({
+        fields: [
+            {from: "json.logName", to: "log.logger"},
+            {from: "json.resource.labels.subnetwork_name", to: "network.name"}
+        ],
+        ignore_missing: true
+    }));
+
+    // Firewall logs are structured so the LogEntry includes a jsonPayload field.
+    // https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry
+    // The LogEntry's jsonPayload is moved to the json field. The jsonPayload
+    // contains the structured VPC flow log fields.
+    builder.Add("convertLogEntry", new processor.Convert({
+        fields: [
+            {from: "json.jsonPayload", to: "json"},
+        ],
+        mode: "rename"
+    }));
+
+    builder.Add("addOutcome", makeMapper({
+        from: "json.disposition",
+        to: "event.outcome",
+        mappings: {
+            ALLOWED: "allow",
+            DENIED: "deny"
+        },
+        default: "unknown"
```

Comment:
Please note that `event.outcome` is still reserved.

The current thinking is that expected values in this field (if present) should be: "success" or "failure". Finer grained information such as "allow" or "deny" should be in another place, perhaps in `event.action` or a custom field.

As usual, populating a reserved field signs you up to having to do a breaking change later.

Reply:
cc @MikePaquette